### PR TITLE
Ap, LinearSVM, OGD as estimators

### PR DESCRIPTION
**Pull request by sfilipi**
Converting AveragePerceptron, LinearSVM and ODG trainers to estimators.

### CODE REVIEW

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedLinear.cs [hunk 1]
```diff
@@ -52,9 +53,10 @@ public abstract class AveragedLinearArguments : OnlineLinearArguments
         public Float AveragedTolerance = (Float)1e-2;
     }
 
-    public abstract class AveragedLinearTrainer<TArguments, TPredictor> : OnlineLinearTrainer<TArguments, TPredictor>
+    public abstract class AveragedLinearTrainer<TArguments, TTransformer, TModel> : OnlineLinearTrainer<TArguments, TTransformer, TModel>
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Online/AveragedLinear.cs]: >TArguments [](start = 48, length = 10)

I will remove this on the next iteration, together with updating the tests.

File: src/Microsoft.ML.Core/Data/IEstimator.cs [hunk 2]
```diff
@@ -237,7 +237,7 @@ public interface IEstimator<out TTransformer>
         /// <summary>
         /// Train and return a transformer.
         /// </summary>
-        TTransformer Fit(IDataView input);
+        TTransformer Fit(IDataView input, IDataView validationData = null, IPredictor initialPredictor = null);
```
Comment (TomFinley) [inline src/Microsoft.ML.Core/Data/IEstimator.cs]: > IDataView validationData = null, IPredictor initialPredictor = null) [](start = 41, length = 69)

I am not certain I welcome this move. It seems like it is repeating the mistake of `IIncrementalValidatingTrainer`, which would be bad enough if it was just for trainers alone but it seems to be for absolutely every estimator.

File: test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs [hunk 3]
```diff
@@ -37,8 +37,8 @@ public void New_TrainWithInitialPredictor()
                 var firstModel = trainer.Fit(trainData);
 
                 // Train the second predictor on the same data.
-                var secondTrainer = new MyAveragedPerceptron(env, new AveragedPerceptronTrainer.Arguments(), "Features", "Label");
-                var finalModel = secondTrainer.Train(trainData, firstModel.Model);
+                var secondTrainer = new AveragedPerceptronTrainer(env, new AveragedPerceptronTrainer.Arguments());
+                var finalModel = secondTrainer.Fit(trainData, initialPredictor: firstModel.Model);
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs]: >Fit [](start = 47, length = 3)

@tfinley@gmail.com if Fit() should not take a validation set, and an initial model, for this example/case here, would we still call Train()?


// Train the second predictor on the same data.
                var secondTrainer = new AveragedPerceptronTrainer(env, new AveragedPerceptronTrainer.Arguments());

                var trainRoles = new RoleMappedData(trainData, label: "Label", feature: "Features");
                var finalModel = secondTrainer.Train(new TrainContext(trainRoles, initialPredictor: firstModel.Model));

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs [hunk 4]
```diff
@@ -29,14 +30,15 @@ namespace Microsoft.ML.Runtime.Learners
...
-    public sealed class AveragedPerceptronTrainer :
-        AveragedLinearTrainer<AveragedPerceptronTrainer.Arguments, LinearBinaryPredictor>
+    public sealed class AveragedPerceptronTrainer : AveragedLinearTrainer<BinaryPredictionTransformer<LinearBinaryPredictor> , LinearBinaryPredictor>
     {
         public const string LoadNameValue = "AveragedPerceptron";
         internal const string UserNameValue = "Averaged Perceptron";
         internal const string ShortName = "ap";
         internal const string Summary = "Averaged Perceptron Binary Classifier.";
 
+        internal new readonly Arguments Args;
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs]: >internal [](start = 8, length = 8)

this should truly be private, but our analyzer wants the private properties to be lowercased. This i should change or add a separate rule for 'private new'. Thoughts?

File: test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs [hunk 5]
```diff
@@ -37,8 +38,10 @@ public void New_TrainWithInitialPredictor()
                 var firstModel = trainer.Fit(trainData);
 
                 // Train the second predictor on the same data.
-                var secondTrainer = new MyAveragedPerceptron(env, new AveragedPerceptronTrainer.Arguments(), "Features", "Label");
-                var finalModel = secondTrainer.Train(trainData, firstModel.Model);
+                var secondTrainer = new AveragedPerceptronTrainer(env, new AveragedPerceptronTrainer.Arguments());
+
+                var trainRoles = new RoleMappedData(trainData, label: "Label", feature: "Features");
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs]: >RoleMappedData [](start = 37, length = 14)

wait, wait, RoleMappedData should not survive :)

File: test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs (hunk 3 above)
Comment (Zruty0) [reply test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs]: In Pigsty land, I'm not sure how this can be accommodated (or rather should it?)

In Dynamic land, yes, we still call Train, just not with the TrainContext and RoleMappedData

---
In reply to: [215809082](https://github.com/dotnet/machinelearning/pull/849#discussion_r215809082) [](ancestors = 215809082)

File: src/Microsoft.ML.Core/Data/IEstimator.cs (hunk 2 above)
Comment (Zruty0) [reply src/Microsoft.ML.Core/Data/IEstimator.cs]: I agree

---
In reply to: [215803874](https://github.com/dotnet/machinelearning/pull/849#discussion_r215803874) [](ancestors = 215803874)

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs (hunk 4 above)
Comment (Zruty0) [reply src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs]: why not just `private readonly Arguments _args` ?

---
In reply to: [215814336](https://github.com/dotnet/machinelearning/pull/849#discussion_r215814336) [](ancestors = 215814336)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs [hunk 6]
```diff
@@ -25,10 +25,11 @@
 
 namespace Microsoft.ML.Runtime.Learners
 {
+    using Microsoft.ML.Core.Data;
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: >using [](start = 4, length = 5)

consolidate

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs [hunk 7]
```diff
@@ -25,10 +25,11 @@
...
     using TPredictor = LinearRegressionPredictor;
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: >TPredictor [](start = 10, length = 10)

let's remove this

**Comment by Zruty0** [comment]
>     using Float = System.Single;

Remove this please

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLinear.cs:19 in 0c176aa. [](commit_id = 0c176aab5bc601689414a1cc2fb03acba1750eeb, deletion_comment = False)

File: test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs (hunk 3 above)
Comment (TomFinley) [reply test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs]: > In Pigsty land, I'm not sure how this can be accommodated (or rather should it?)

Almost certainly, but unless I'm mistaken I don't *think* changes to `IEstimator`, or any of these core interfaces, will be necessary, or indeed even slightly helpful. Indeed, I might be missing something, but I'm a little confused about where we imagine the source of the problem is? We've posited the existence of some hypothetical `ITransformer Train(IDataView trainData, ...)` method, so why do we suppose a `Transformer<TIn, TOut> Train<TIn, TOut>(DataView<TIn> trainData, ...)` is impossible? Perhaps there's some subtlety I've missed.

---
In reply to: [215821268](https://github.com/dotnet/machinelearning/pull/849#discussion_r215821268) [](ancestors = 215821268,215809082)

Comment (TomFinley) [reply test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs]: I'm not certain whether that's in scope for this PR. Maybe it is, I just don't know.

---
In reply to: [215840658](https://github.com/dotnet/machinelearning/pull/849#discussion_r215840658) [](ancestors = 215840658,215821268,215809082)

File: test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs (hunk 5 above)
Comment (sfilipi) [reply test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs]: What is the preferred  way of passing the initialPredictor for online learners if neither Train or Fit takes them in as arguments? Some Transform() ?

---
In reply to: [215821958](https://github.com/dotnet/machinelearning/pull/849#discussion_r215821958) [](ancestors = 215821958,215820969)

Comment (sfilipi) [reply test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs]: @Pete Luferenko i added a second TrainContext that contains IDataView for the Training, Validation data, and another Train Method that takes this second TrainContext but that proliferates everywhere in an ugly way. 

I think we should keep the RoleMappedData instead of this spawn, and once it goes away, TrainContext will get cleaned up and so will this example. Objections?

---
In reply to: [215848474](https://github.com/dotnet/machinelearning/pull/849#discussion_r215848474) [](ancestors = 215848474,215821958,215820969)

Comment (Zruty0) [reply test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithInitialPredictor.cs]: that's ok, let's create an issue for that though.

---
In reply to: [216079699](https://github.com/dotnet/machinelearning/pull/849#discussion_r216079699) [](ancestors = 216079699,215848474,215821958,215820969)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLinear.cs [hunk 8]
```diff
@@ -2,8 +2,11 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using Float = System.Single;
+
```
Comment (TomFinley) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLinear.cs]: I think @Zruty0 meant remove it altogether, but that's all right, we can do a sweep for this later.
